master/buildbot/reporters/github.py

```diff
@@ -0,0 +1,138 @@
+
+
+class GithubStatusPush(http.HttpStatusPushBase):
+    name = "GithubStatusPush"
+    neededDetails = dict(wantProperties=True)
+
+    @defer.inlineCallbacks
+    def reconfigService(self, token,
+                        startDescription=None, endDescription=None,
+                        context=None, baseURL=None, verbose=False, **kwargs):
+        yield http.HttpStatusPushBase.reconfigService(self, **kwargs)
+
+        self.context = context or Interpolate('buildbot/%(prop:buildername)s')
+        self.startDescription = startDescription or 'Build started.'
+        self.endDescription = endDescription or 'Build done.'
+        if baseURL is None:
+            baseURL = HOSTED_BASE_URL
+        if baseURL.endswith('/'):
+            baseURL = baseURL[:-1]
+        self.baseURL = baseURL
+        self.session.headers.update({'Authorization': 'token ' + token})
+        self.verbose = verbose
+
+    def createStatus(self,
+                     repo_user, repo_name, sha, state, target_url=None,
+                     description=None, context=None):
+        """
+        :param sha: Full sha to create the status for.
+        :param state: one of the following 'pending', 'success', 'error'
+                      or 'failure'.
+        :param target_url: Target url to associate with this status.
+        :param description: Short description of the status.
+        :return: A defered with the result from GitHub.
+
+        This code comes from txgithub by @tomprince.
+        txgithub is based on twisted's webclient agent, which is much less reliable and featureful
+        as txrequest (support for proxy, connection pool, keep alive, retry, etc)
+        """
+        payload = {'state': state}
+
+        if description is not None:
+            payload['description'] = description
+
+        if target_url is not None:
+            payload['target_url'] = target_url
+
+        if context is not None:
+            payload['context'] = context
+
+        return self.session.post("/".join(
+            [self.baseURL, 'repos', repo_user, repo_name, 'statuses', sha]),
+            payload)
+
+    @defer.inlineCallbacks
+    def send(self, build):
+        props = Properties.fromDict(build['properties'])
+
+        if build['complete']:
+            state = {
+                SUCCESS: 'success',
+                WARNINGS: 'success',
+                FAILURE: 'failure',
+                SKIPPED: 'success',
+                EXCEPTION: 'error',
+                RETRY: 'pending',
+                CANCELLED: 'error'
+            }.get(build['results'], 'error')
+            description = yield props.render(self.endDescription)
+        else:
+            state = 'pending'
+            description = yield props.render(self.startDescription)
+
+        context = yield props.render(self.context)
+
+        sourcestamps = build['buildset']['sourcestamps']
+        project = sourcestamps[0]['project']
+
+        if not project:
```

Comment:
i don't know if it makes a lot of sense, but why not change the condition and swap the branches?